code/povelikin-rostislav/src/test/java/ru/unn/agile/redblacktree/RedBlackTreeTest.java

```diff
@@ -0,0 +1,230 @@
+        tree.insert(Integer.MIN_VALUE);
+        tree.remove(Integer.MIN_VALUE);
+
+        assertFalse(tree.find(Integer.MIN_VALUE));
+    }
+
+    @Test
+    public void removeCornerCaseMax() {
+        tree.insert(Integer.MAX_VALUE);
+        tree.remove(Integer.MAX_VALUE);
+
+        assertFalse(tree.find(Integer.MAX_VALUE));
+    }
+
+    @Test
+    public void insertSameNumberIncrementsSize() {
+        int number = 42;
+        tree.insert(number);
+        tree.insert(number);
+        assertEquals(2, tree.getSize());
+    }
+
+    @Test
+    public void insertRandomNumberSequence() {
+        for (int counter = 0; counter < 1000000; ++counter) {
+            int randomNumber = random.nextInt(1024);
+            tree.insert(randomNumber);
+            assertTrue(tree.find(randomNumber));
+        }
+    }
+
+    @Test
+    public void removeRandomNumberSequence() {
+        for (int counter = 0; counter < 1000000; ++counter) {
+            int insertRandomNumber = random.nextInt(1024);
+            tree.insert(insertRandomNumber);
+            int removeRandomNumber = random.nextInt(1024);
+            final boolean found = tree.find(removeRandomNumber);
+            assertEquals(found, tree.remove(removeRandomNumber));
+        }
+    }
+
+    @Test
+    public void findRandomNumberSequence() {
+        Set<Integer> set = new HashSet<>();
+
+        for (int counter = 0; counter < 1000000; ++counter) {
+            int insertRandomNumber = random.nextInt(1024);
+            set.add(insertRandomNumber);
+            tree.insert(insertRandomNumber);
+            int findRandomNumber = random.nextInt(1024);
+            assertEquals(set.contains(findRandomNumber), tree.find(findRandomNumber));
+        }
+    }
+
+    @Test
+    public void removeDoesNotRemoveAllOccurrences() {
+        int number = 42;
+        tree.insert(number);
+        tree.insert(number);
+
+        tree.remove(number);
+
+        assertTrue(tree.find(number));
+    }
+
+    @Test
+    public void removeRemoves() {
+        int number = 42;
+        tree.insert(number);
+
+        tree.remove(number);
+
+        assertFalse(tree.find(number));
+    }
+
+    @Test
+    public void canFindEachInserted() {
+        int threshold = 88;
+        int counter = 0;
+        while (counter < threshold) {
+            ++counter;
+            tree.insert(counter);
+            assertTrue(tree.find(counter));
+        }
+    }
+
+    @Test
+    public void cannotFindNotInserted() {
+        int threshold = 88;
+        int counter = 0;
+        while (counter < threshold) {
+            assertFalse(tree.find(++counter));
+        }
+    }
+
+    @Test
+    public void removeReturnsFalseIfNotFound() {
+        assertFalse(tree.remove(42));
+    }
+
+    @Test
+    public void removeReturnsTrueIfFound() {
+        int expected = 42;
+
+        tree.insert(expected);
+
+        assertTrue(tree.remove(expected));
+    }
+
+    @Test
+    public void removeChangesSizeIfFound() {
```

Comment:
`removeChangesSizeIfFound()` and `removeDoesNotChangeSizeIfNotFound()` operate with many values. Why we cannot just work with single one?

Reply:
Because the tree should support the property we test on each iteration > 2. I just chose one number > 2